R/union.R

```diff
@@ -225,23 +258,25 @@ union <- function(spc=list(), method='all', na.rm=TRUE, drop.spatial=FALSE) {
     stop('non-unique profile IDs detected')
   }
 
-  ## make SPC from pieces
-  res <- SoilProfileCollection(idcol = new.pID,
-                               hzidcol = new.hzID,
-                               hzdesgncol = new.hzdesgn,
-                               hztexclcol = new.hztexcl,
-                               depthcols = new.hzd,
-                               metadata = new.metadata,
-                               horizons = .as.data.frame.aqp(o.h, o.df.class),
-                               site = .as.data.frame.aqp(o.s, o.df.class),
-                               sp = o.sp,
-                               diagnostic = .as.data.frame.aqp(o.d, o.df.class),
-                               restrictions = .as.data.frame.aqp(o.r, o.df.class))
-
-  ## reset horizon IDs (if using default hzID)
-  if ('hzID' %in% horizonNames(res))
-    res$hzID <- as.character(1:nrow(res))
+  # take combined horizon data, convert to target df subclass
+  res <- .as.data.frame.aqp(o.h, o.df.class)
+
+  ## reset horizon IDs (if the default column hzID is presnt)
+  if ('hzID' %in% colnames(res))
+    res$hzID <- NULL
+
+  # rebuild SPC using safe constructors
+  depths(res) <- formula(sprintf("%s ~ %s + %s", new.pID, new.hzd[1], new.hzd[2]))
+  site(res) <- .as.data.frame.aqp(o.s, o.df.class)
```

Comment:
This is the extra complexity that I am talking about: is this required every time `site(x) <- ` is invoked?

Reply:
This is `union`... No.

Reply:
Also, .as.data.frame.aqp only checks the class. we have talked about this before. there is no overhead or complexity to speak of,. you read my document that covered .as.data.frame.aqp and .data.frame.j. These are the two features that were added

Reply:
I recall those conversations. To be clear: at SPC construction time (e.g. I take an SPC apart, do something, and put it back together) can I use standard constructors without making special accommodation for `data.table` or `tbl_df` in the general sense? My concern is this: added complexity at SCP construction / deconstruction is something that I want to avoid.